### PR TITLE
[IA-1580] Persistent disk CRUD routes

### PR DESCRIPTION
**Pull request by akarukappadath**
\<your comments for this PR go here\>

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala
Comment (rtitle) [inline core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala]: Not sure `serviceAccount` really makes sense as a disk label.. you don't need a service account to create a disk.

Comment (Qi77Qi) [reply core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala]: what's the purpose of these labels? are they get created as google labels?

File: http/src/main/resources/reference.conf
Comment (rtitle) [inline http/src/main/resources/reference.conf]: From docs the allowed values are 4096 and 16384. Probably use 4096 as a default.

https://cloud.google.com/compute/docs/reference/rest/v1/disks/insert

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/DiskRoutes.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/DiskRoutes.scala]: I think we wanted to use plural routes, so `/api/google/v1/disks`

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala
Comment (akarukappadath) [reply core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala]: @rtitle thanks - meant to leave a comment asking if SAs were relevant for disks. 
@Qi77Qi the disk labels don't get persisted to google (similar to runtimes) - it's just another way to filter disks like runtimes

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: Cool, so handling this will be the work for https://broadworkbench.atlassian.net/browse/IA-1757 (cc @gpcarr)

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/DiskServiceInterp.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/DiskServiceInterp.scala]: Don't think we need to get a pet to create a disk

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/DiskServiceInterp.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/DiskServiceInterp.scala]: We may want to enqueue _before_ persisting the DB record. Reason is we only want to save the record if enqueuing succeeds.

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala
Comment (rtitle) [inline core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala:38]: Wondering if we'll want a `PreCreating` status similar to runtimes..

**Comment by rtitle** [comment]
I know this is WIP but added a few suggestions. Looks great overall! I'll re-review when ready but I think this will be close to merging once tests are added 🚀

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala
Comment (Qi77Qi) [reply core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala]: in that case...do we still need labels like `diskName, googleProject, creator`? We can filter disks by these already without labels I think?

Comment (rtitle) [reply core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala]: Still think it's useful even if you can filter by googleProject, diskName in other ways. For instance Terra UI lists runtimes like this:
```
https://notebooks.firecloud.org/api/google/v1/runtimes?saturnAutoCreated=true&googleProject=broad-firecloud-dsde&creator=rtitle%40broadinstitute.org
```
They could change to this: 
```
https://notebooks.firecloud.org/api/google/v1/runtimes/broad-firecloud-dsde?saturnAutoCreated=true&creator=rtitle%40broadinstitute.org
```
but either way the DB query is pretty fast. It's nice to support both options.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/DiskServiceInterp.scala
Comment (akarukappadath) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/DiskServiceInterp.scala]: We need to get `diskId` by saving the disk before we enqueue it, so I don't think I can flip the order

**Comment by akarukappadath** [comment]
jenkins restart

**Review by rtitle — APPROVED** [review]
Looks good!

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/DiskServiceInterp.scala
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/DiskServiceInterp.scala]: in my other PR, we're introducing a `pre` status to represent the period when the even is not being pushed to pubsub yet...I would either wait for my other PR (cuz there're some extra stuff needed for this to happen, or at least add a comment to address this)

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/LeoRoutes.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/LeoRoutes.scala]: I think we're slowly deprecating this file...We don't really need to add stuff this file anymore...Can we move the validation to `DisksRoutes`'s companion object?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: make this private...we shouldn't use this reader outside this file

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/PersistentDiskConfig.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/PersistentDiskConfig.scala]: minor, add `final`

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/DiskServiceInterp.scala
Comment (akarukappadath) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/DiskServiceInterp.scala]: Gotcha, I can hold off for now and add a `precreating` status to disks before merging.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: can you add something like `throw new NotImplementedError("disk events are not implemented yet")`

**Review by Qi77Qi — COMMENTED** [review]
are you going to add swagger doc as well? or at least some doc describing the APIs would be nice

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala
Comment (Qi77Qi) [inline core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala]: can you use `sealerate.values.map(obj => obj.googleString -> obj` for `stringToDiskType`

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala
Comment (akarukappadath) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: should I go ahead and make all of the readers here private then?

Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: nah...I did that in my other PR...

Comment (akarukappadath) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: making this private made it inaccessible for the configs in `CommonTestData` - did you move those around in your PR as well?

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala
Comment (akarukappadath) [reply core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/diskModels.scala]: Decided to revert back to Enum and use the given `withNameOption` method instead of writing our own here

File: http/src/main/resources/swagger/api-docs.yaml
Comment (Qi77Qi) [inline http/src/main/resources/swagger/api-docs.yaml]: can you name this to `listPersistentDiskResponse`...I'd like to prevent us from reusing this model for other routes so that routes can evolve independent of each other and we won't accidentally update contract for one route but end up updating multiple routes.

File: http/src/main/resources/swagger/api-docs.yaml
Comment (Qi77Qi) [inline http/src/main/resources/swagger/api-docs.yaml]: same comment, can this be `GetPersistentDiskResponse`

File: http/src/main/resources/swagger/api-docs.yaml
Comment (Qi77Qi) [inline http/src/main/resources/swagger/api-docs.yaml:657]: if the disk is being used by a running VM, will we be able to update the disk?

File: http/src/main/resources/swagger/api-docs.yaml
Comment (Qi77Qi) [inline http/src/main/resources/swagger/api-docs.yaml]: is the size increase intentional?

File: http/src/main/resources/swagger/api-docs.yaml
Comment (Qi77Qi) [inline http/src/main/resources/swagger/api-docs.yaml:2129]: are we not return `error` for  GetPersistentDisk route?...I don't think `labels` is required...
Not all these fields are applicable to `listPersistentDiskResponse` I think.

File: http/src/main/resources/swagger/api-docs.yaml
Comment (Qi77Qi) [inline http/src/main/resources/swagger/api-docs.yaml]: I think terra ui preferred these names remain the same...like `size`, `diskType`, `blockSize`...

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/config/Config.scala]: yea that's exactly the point...we shouldn't use it in `CommonTestData`...you can use `Config.persistentDisk` (whatever name you give it) in test the same way you do in app code...That way, our unit tests will actually be testing the Config code as well instead of doing its own thing in test..see https://github.com/DataBiosphere/leonardo/pull/1395/files#diff-4101c04c4a7015e058bf48267899df0bR92

File: http/src/main/resources/swagger/api-docs.yaml
Comment (rtitle) [reply http/src/main/resources/swagger/api-docs.yaml]: +1

File: http/src/main/resources/swagger/api-docs.yaml
Comment (rtitle) [reply http/src/main/resources/swagger/api-docs.yaml]: +1

File: http/src/main/resources/swagger/api-docs.yaml
Comment (rtitle) [reply http/src/main/resources/swagger/api-docs.yaml:657]: You can resize a disk while it's attached to a running VM.

My PR involves adding validation/documentation about what operations are/aren't allowed with disks & VMs.

File: http/src/main/resources/swagger/api-docs.yaml
Comment (rtitle) [reply http/src/main/resources/swagger/api-docs.yaml]: Hm.. let's keep `CreateRuntimeRequest` unchanged in this PR, I am updating it in my PR

File: http/src/main/resources/swagger/api-docs.yaml
Comment (rtitle) [inline http/src/main/resources/swagger/api-docs.yaml]: Don't think this example is right, there is no `runtimeConfig` or `cloudService`

File: http/src/main/resources/swagger/api-docs.yaml
Comment (rtitle) [reply http/src/main/resources/swagger/api-docs.yaml]: +1

File: http/src/main/resources/swagger/api-docs.yaml
Comment (rtitle) [inline http/src/main/resources/swagger/api-docs.yaml:2167]: Hm.. was spacing wrong here? I touched this recently, would be good to make sure it still displays correctly

**Comment by rtitle** [comment]
Few more comments in swagger docs.. rest LGTM. Also this PR has some conflicts 😑

File: http/src/main/resources/swagger/api-docs.yaml
Comment (akarukappadath) [reply http/src/main/resources/swagger/api-docs.yaml]: Makes sense. To clarify, this is only in the swagger? Or should that be a type of its own in `DiskRoutes.scala`?

File: http/src/main/resources/swagger/api-docs.yaml
Comment (akarukappadath) [reply http/src/main/resources/swagger/api-docs.yaml:2129]: We don't have errors for PDs yet, I think we were holding off until we have a better idea of what kinds of errors can come up. @rtitle would that be after some of the back Leo work?

File: http/src/main/resources/swagger/api-docs.yaml
Comment (Qi77Qi) [reply http/src/main/resources/swagger/api-docs.yaml]: I think both..We should have dedicated response type in code as well

File: http/src/main/resources/swagger/api-docs.yaml
Comment (rtitle) [reply http/src/main/resources/swagger/api-docs.yaml:2129]: Yeah we haven't added any disk errors yet… let's hold off till another PR, I don't want to keep increasing scope of this one.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/DiskRoutes.scala
Comment (akarukappadath) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/DiskRoutes.scala:276]: As of the moment, these are the same as the `PersistentDisk` type. Once we include `errors` in PDs, I think that will go in the `get` response and not the `list` response. Other potential fields we could remove from the `list` response are `googleId`, `samResourceId`, and `labels` - let me know what you think

Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/DiskRoutes.scala:276]: sounds good to me...are you removing the fields you mentioned? or you're just saying that can be done in the future? (I'd prefer just removing them if no particular reason to keep..either this PR or follow up PRs)

File: http/src/main/resources/swagger/api-docs.yaml
Comment (Qi77Qi) [inline http/src/main/resources/swagger/api-docs.yaml]: I thought you were going to remove this `samResourceId`?

**Review by Qi77Qi — APPROVED** [review]

File: http/src/main/resources/swagger/api-docs.yaml
Comment (akarukappadath) [reply http/src/main/resources/swagger/api-docs.yaml]: It turns out we use it ourselves to check whether or not the disk is visible for the user before `list`ing, so I think it makes sense to keep in there

**Pull request merged**
